src/base/cub_ddl_log.c

```diff
@@ -0,0 +1,928 @@
+struct t_ddl_audit_handle
+{
+  char db_name[DB_MAX_IDENTIFIER_LENGTH];
+  char user_name[DB_MAX_USER_LENGTH];
+  char app_name[DDL_APP_NAME_LEN];
+  char br_name[DDL_APP_NAME_LEN];
+  int br_index;
+  char ip_addr[16];
+  int pid;
+  char *sql_text;
+  char stmt_type;
+  char execute_start_time[DDL_TIME_LEN];
+  char elapsed_time[DDL_TIME_LEN];
+  int file_line_number;
+  int err_code;
+  char *err_msg;
+  //char result;
+  char msg[DDL_LOG_MSG];
+  char file_name[DDL_LOG_PATH_MAX];
+  char schema_file[DDL_LOG_PATH_MAX];
+  char log_type;
+  char loaddb_file_type;
+  char log_filepath[DDL_LOG_PATH_MAX];
+};
+static T_DDL_AUDIT_HANDLE *ddl_audit_handle = NULL;
+
+static int cub_make_ddl_log_filename (char *filename_buf, size_t buf_size, const char *app_name);
+static int cub_make_schema_file_name (const char *file_full_path, char *dest_path, size_t buf_size);
+static void cub_ddl_log_backup (const char *path);
+#if defined(WINDOWS)
+static void unix_style_path (char *path);
+#endif /* WINDOWS */
+static int cub_create_dir_log (const char *new_dir);
+static int cub_create_log_mgs (char *msg);
+static int cub_get_current_time (char *buf, size_t size);
+static int cub_file_copy (char *src_file, char *dest_file);
+static void cub_remove_char (char *string, char ch);
+static int cub_is_ddl_type (int node);
+static FILE *cub_ddl_log_open (char *app_name);
+static int cub_get_time_string (char *buf, struct timeval *time_val);
+static FILE *cub_fopen_and_lock (const char *path, const char *mode);
+
+void
+cub_ddl_log_init ()
+{
+  if (ddl_audit_handle != NULL)
+    {
+      FREE_MEM (ddl_audit_handle->sql_text);
+      FREE_MEM (ddl_audit_handle->err_msg);
+      FREE_MEM (ddl_audit_handle);
+    }
+
+  ddl_audit_handle = (T_DDL_AUDIT_HANDLE *) MALLOC (sizeof (T_DDL_AUDIT_HANDLE));
+
+  memset (ddl_audit_handle, 0x00, sizeof (T_DDL_AUDIT_HANDLE));
+
+  ddl_audit_handle->pid = 0;
+  ddl_audit_handle->stmt_type = -1;
+  ddl_audit_handle->br_index = 0;
+  ddl_audit_handle->file_line_number = 0;
+  ddl_audit_handle->err_code = 0;
+  ddl_audit_handle->log_type = DDL_LOG_ELAPSED_TIME;
+  ddl_audit_handle->loaddb_file_type = LOADDB_FILE_TYPE_NONE;
+}
+
+void
+cub_ddl_log_free ()
+{
+
+  FREE_MEM (ddl_audit_handle->sql_text);
+  FREE_MEM (ddl_audit_handle->err_msg);
+
+  memset (ddl_audit_handle, 0x00, sizeof (T_DDL_AUDIT_HANDLE));
+
+  ddl_audit_handle->pid = 0;
+  ddl_audit_handle->stmt_type = -1;
+  ddl_audit_handle->file_line_number = 0;
+  ddl_audit_handle->err_code = 0;
+  ddl_audit_handle->log_type = DDL_LOG_ELAPSED_TIME;
+  ddl_audit_handle->loaddb_file_type = LOADDB_FILE_TYPE_NONE;
+}
+
+void
+cub_ddl_log_destroy ()
+{
+  if (ddl_audit_handle != NULL)
+    {
+      FREE_MEM (ddl_audit_handle->sql_text);
+      FREE_MEM (ddl_audit_handle->err_msg);
+      FREE_MEM (ddl_audit_handle);
+    }
+}
+
+void
+cub_ddl_log_app_name (const char *app_name)
+{
+  snprintf (ddl_audit_handle->app_name, sizeof (ddl_audit_handle->app_name), app_name);
```

Comment:
You assume that ddl_audit_handle is properly allocated memory. Is it unnecessary to check whether the ddl audit handle is NULL? For example:

```
if (ddl_audit_handle)
 {
    snprintf (ddl_audit_handle->app_name, sizeof (ddl_audit_handle->app_name), app_name);
 }
```